vault/root_generation.go

```diff
@@ -0,0 +1,300 @@
+		fingerprints, err := pgpkeys.GetFingerprints([]string{pgpKey}, nil)
+		if err != nil {
+			return fmt.Errorf("error parsing PGP key: %s", err)
+		}
+		if len(fingerprints) != 1 || fingerprints[0] == "" {
+			return fmt.Errorf("could not acquire PGP key entity")
+		}
+		fingerprint = fingerprints[0]
+
+	default:
+		return fmt.Errorf("unreachable condition")
+	}
+
+	c.stateLock.RLock()
+	defer c.stateLock.RUnlock()
+	if c.sealed {
+		return ErrSealed
+	}
+	if c.standby {
+		return ErrStandby
+	}
+
+	c.rootGenerationLock.Lock()
+	defer c.rootGenerationLock.Unlock()
+
+	// Prevent multiple concurrent root generations
+	if c.rootGenerationConfig != nil {
+		return fmt.Errorf("root generation already in progress")
+	}
+
+	// Copy the configuration
+	generationNonce, err := uuid.GenerateUUID()
+	if err != nil {
+		return err
+	}
+
+	c.rootGenerationConfig = &RootGenerationConfig{
+		Nonce:          generationNonce,
+		OTP:            otp,
+		PGPKey:         pgpKey,
+		PGPFingerprint: fingerprint,
+	}
+
+	c.logger.Printf("[INFO] core: root generation initialized (nonce: %s)",
+		c.rootGenerationConfig.Nonce)
+	return nil
+}
+
+// RootGenerationUpdate is used to provide a new key part
+func (c *Core) RootGenerationUpdate(key []byte, nonce string) (*RootGenerationResult, error) {
+	// Verify the key length
+	min, max := c.barrier.KeyLength()
+	max += shamir.ShareOverhead
+	if len(key) < min {
+		return nil, &ErrInvalidKey{fmt.Sprintf("key is shorter than minimum %d bytes", min)}
+	}
+	if len(key) > max {
+		return nil, &ErrInvalidKey{fmt.Sprintf("key is longer than maximum %d bytes", max)}
+	}
+
+	// Get the seal configuration
+	config, err := c.SealConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure the barrier is initialized
+	if config == nil {
+		return nil, ErrNotInit
+	}
+
+	// Ensure we are already unsealed
+	c.stateLock.RLock()
+	defer c.stateLock.RUnlock()
+	if c.sealed {
+		return nil, ErrSealed
+	}
+	if c.standby {
+		return nil, ErrStandby
+	}
+
+	c.rootGenerationLock.Lock()
+	defer c.rootGenerationLock.Unlock()
+
+	// Ensure a rootGeneration is in progress
+	if c.rootGenerationConfig == nil {
+		return nil, fmt.Errorf("no root generation in progress")
+	}
+
+	if nonce != c.rootGenerationConfig.Nonce {
+		return nil, fmt.Errorf("incorrect nonce supplied; nonce for this root generation operation is %s", c.rootGenerationConfig.Nonce)
+	}
+
+	// Check if we already have this piece
+	for _, existing := range c.rootGenerationProgress {
+		if bytes.Equal(existing, key) {
+			return nil, nil
+		}
+	}
+
+	// Store this key
+	c.rootGenerationProgress = append(c.rootGenerationProgress, key)
+	progress := len(c.rootGenerationProgress)
+
+	// Check if we don't have enough keys to unlock
+	if len(c.rootGenerationProgress) < config.SecretThreshold {
+		c.logger.Printf("[DEBUG] core: cannot generate root, have %d of %d keys",
+			progress, config.SecretThreshold)
+		return &RootGenerationResult{
+			Progress:       progress,
```

Comment:
That will be automatically included upon return -- this gets wrapped into a larger response object.